server/app/models/interaction.rb

```diff
@@ -51,15 +51,29 @@ def interaction_score(known_drug_partners_per_gene = nil, known_gene_partners_pe
     end
   end
 
-  def calculate_interaction_score(known_drug_partners_per_gene = nil, known_gene_partners_per_drug = nil)
-    known_drug_partners_per_gene = Interaction.group(:gene_id).count if known_drug_partners_per_gene.nil?
-    average_known_drug_partners_per_gene = known_drug_partners_per_gene.values.sum / known_drug_partners_per_gene.values.size.to_f
-    known_gene_partners_per_drug = Interaction.group(:drug_id).count if known_gene_partners_per_drug.nil?
-    average_known_gene_partners_per_drug = known_gene_partners_per_drug.values.sum / known_gene_partners_per_drug.values.size.to_f
-    known_drug_partners_for_interaction_gene = known_drug_partners_per_gene[self.gene_id]
-    known_gene_partners_for_interaction_drug = known_gene_partners_per_drug[self.drug_id]
+  def calculate_interaction_score(drug_partners_per_gene = nil, gene_partners_per_drug = nil, update = false)
+    drug_partners_per_gene = Interaction.group(:gene_id).count if drug_partners_per_gene.nil?
+    avg_drug_partners_per_gene = drug_partners_per_gene.values.sum / drug_partners_per_gene.values.size.to_f
+    drug_partners_for_this_gene = drug_partners_per_gene[gene_id]
 
-    (self.publications.count + self.sources.count) * average_known_gene_partners_per_drug/known_gene_partners_for_interaction_drug * average_known_drug_partners_per_gene/known_drug_partners_for_interaction_gene
+    gene_partners_per_drug = Interaction.group(:drug_id).count if gene_partners_per_drug.nil?
+    avg_gene_partners_per_drug = gene_partners_per_drug.values.sum / gene_partners_per_drug.values.size.to_f
+    gene_partners_for_this_drug = gene_partners_per_drug[drug_id]
+
+    drug_specificity = avg_gene_partners_per_drug / gene_partners_for_this_drug
+    gene_specificity = avg_drug_partners_per_gene / drug_partners_for_this_gene
+    evidence_score = publications.count + sources.count
+    interaction_score = evidence_score * drug_specificity * gene_specificity
+
+    if update
+      self.score = interaction_score
+      self.drug_specificity = drug_specificity
+      self.gene_specificity = gene_specificity
+      self.evidence_score = evidence_score
+      save!
+    end
+
+    interaction_score
   end
```

Comment:
this looks right, at least as far as https://dgidb.org/about#interaction-scores is concerned. I would like to get @acoffman just to confirm that the ruby isn't doing anything unexpected here

Reply:
This looks correct to me on first glance. Only "gotcha" I could think of is defaulting to integer division between two ints but it looks like that's covered by the `.to_f` calls.